fog/sample-paykit/src/cached_tx_data/mod.rs

```diff
@@ -390,6 +460,127 @@ impl CachedTxData {
         Ok(())
     }
 
+    /// Determines the new missed block ranges given the blocks that are
+    /// retrieved in the BlockData object.
+    fn calculate_updated_missed_block_ranges(
+        missed_block_ranges: &[common::BlockRange],
+        block_data: &[ledger::BlockData],
+    ) -> Vec<common::BlockRange> {
+        // Transforms the missed BlockRanges into a set of missed block
+        // indices.
+        let mut missed_block_indices =
+            CachedTxData::create_missed_block_indices(missed_block_ranges);
+
+        for block_datum in block_data {
+            missed_block_indices.remove(&block_datum.index);
+        }
+
+        // Sort the indices in order to create the fewest amount of block
+        // ranges possible in the iteration step below.
+        let mut missed_block_indices_sorted: Vec<u64> =
+            missed_block_indices.into_iter().collect::<Vec<_>>();
+        missed_block_indices_sorted.sort_unstable();
+
+        let mut updated_missed_block_ranges: Vec<common::BlockRange> = Vec::new();
+
+        if missed_block_indices_sorted.is_empty() {
+            return updated_missed_block_ranges;
+        }
+
+        if missed_block_indices_sorted.len() == 1 {
+            let first_block_index = missed_block_indices_sorted[0];
+            updated_missed_block_ranges.push(common::BlockRange {
+                start_block: first_block_index,
+                end_block: first_block_index + 1,
+            });
+            return updated_missed_block_ranges;
+        }
+
+        let mut start_block_index = missed_block_indices_sorted[0];
+        let mut i: usize = 0;
+        let mut j: usize = 1;
```

Comment:
it seems like there is an invariant that `j == i + 1` so it would be simpler if we don't have `j`?